python/cuml/dask/cluster/kmeans.py

```diff
@@ -178,24 +173,22 @@ def fit(self, X):
         -------
         self: KMeans model
         """
-        gpu_futures = self.client.sync(extract_ddf_partitions, X)
-
-        worker_to_parts = workers_to_parts(gpu_futures)
 
-        workers = list(map(lambda x: x[0], worker_to_parts.items()))
+        data = MGData.single(X, client=self.client)
+        self.datatype = data.datatype
 
         comms = CommsContext(comms_p2p=False)
-        comms.init(workers=workers)
+        comms.init(workers=data.workers)
 
         key = uuid1()
-        kmeans_fit = [self.client.submit(
-            KMeans._func_fit,
-            comms.sessionId,
-            wf[1],
-            **self.kwargs,
-            workers=[wf[0]],
-            key="%s-%s" % (key, idx))
-            for idx, wf in enumerate(worker_to_parts.items())]
+        kmeans_fit = [self.client.submit(KMeans._func_fit,
```

Comment:
(not for this PR) but this will also be a nice pattern to try to wrap in the future. It is commonly used, but a lot of it (uuid, key generation, passing through datatype and session id) is not algorithm-specific.